geo-traits/src/geometry.rs

```diff
@@ -0,0 +1,87 @@
+use geo_types::{
+    CoordNum, Geometry, GeometryCollection, LineString, MultiLineString, MultiPoint, MultiPolygon,
+    Point, Polygon,
+};
+
+use super::{
+    GeometryCollectionTrait, LineStringTrait, MultiLineStringTrait, MultiPointTrait,
+    MultiPolygonTrait, PointTrait, PolygonTrait,
+};
+
+#[allow(clippy::type_complexity)]
+pub trait GeometryTrait<'a>: Send + Sync {
+    type Point: 'a + PointTrait;
+    type LineString: 'a + LineStringTrait<'a>;
+    type Polygon: 'a + PolygonTrait<'a>;
+    type MultiPoint: 'a + MultiPointTrait<'a>;
+    type MultiLineString: 'a + MultiLineStringTrait<'a>;
+    type MultiPolygon: 'a + MultiPolygonTrait<'a>;
+    type GeometryCollection: 'a + GeometryCollectionTrait<'a>;
+
+    fn as_type(
+        &'a self,
+    ) -> GeometryType<
+        'a,
+        Self::Point,
+        Self::LineString,
+        Self::Polygon,
+        Self::MultiPoint,
+        Self::MultiLineString,
+        Self::MultiPolygon,
+        Self::GeometryCollection,
+    >;
+}
+
+#[derive(Debug)]
+pub enum GeometryType<'a, P, L, Y, MP, ML, MY, GC>
+where
+    P: 'a + PointTrait,
+    L: 'a + LineStringTrait<'a>,
+    Y: 'a + PolygonTrait<'a>,
+    MP: 'a + MultiPointTrait<'a>,
+    ML: 'a + MultiLineStringTrait<'a>,
+    MY: 'a + MultiPolygonTrait<'a>,
+    GC: 'a + GeometryCollectionTrait<'a>,
+{
+    Point(&'a P),
+    LineString(&'a L),
+    Polygon(&'a Y),
+    MultiPoint(&'a MP),
+    MultiLineString(&'a ML),
+    MultiPolygon(&'a MY),
+    GeometryCollection(&'a GC),
```

Comment:
Do you have any thoughts on adding the other types offered in `geo-types`? Namely `Triangle`, `Line`, and `Rect`

Reply:
Definitely something to discuss. I didn't implement them yet because I wasn't sure the pros/cons of how hard that might be for some data structures to implement vs the usefulness in adapting for geo algorithms. If a data structure doesn't natively have a `Rect` type, should it allow a polygon to fill in as a `Rect`?